### PR TITLE
nixos/wireguard: add option to set firewall mark and mtu

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

Add two options to `networking.wireguard.interfaces.<name>`. `mtu` to set the mtu of the interface and `fwMark` to have wireguard tag its packets with the given fwmark.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by maifel-maifel** [comment]
I just realized that I made some fucky wuckies. I will update the commits tomorrow, after I slept, when I'm not likely to make stupid mistakes.
* the commit message of [530c5ef](https://github.com/NixOS/nixpkgs/pull/189176/commits/530c5ef440dba7d2a42d5c10babc38f080f86d31) uses `types.string` instead of `types.str`
* commit [5a90020](https://github.com/NixOS/nixpkgs/pull/189176/commits/5a90020451b7f5d8c0bbc6314cb1f43b4c73a26f) should be using `types.ints.positive` instead of just `types.int`

**Comment by maifel-maifel** [comment]
I read the manual again and decided that it is not more fitting for `networking.wireguard.interfaces.<name>.mtu` to be of type `types.ints.positive` than `types.int` so I left it at the latter. Some Typos were fixed.

File: nixos/modules/services/networking/wireguard.nix [hunk 1]
```diff
@@ -137,6 +137,32 @@ let
         See [documentation](https://www.wireguard.com/netns/).
         '';
       };
+
+      fwMark = mkOption {
+        default = null;
+        type = with types; nullOr string;
+        example = "0x6e6978";
+        description = lib.mdDoc ''Mark all wireguard packets originating from
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/wireguard.nix:145]:
```suggestion
        description = lib.mdDoc ''
          Mark all wireguard packets originating from
```

Comment (maifel-maifel) [reply nixos/modules/services/networking/wireguard.nix:145]: yeah sounds good, should I redo the commits or just press "commit changes" in the github ui?

Comment (maifel-maifel) [reply nixos/modules/services/networking/wireguard.nix]: alright, enough time passed that I decided on my own. some inconsistencies in documentation for some other wireguard options bothered me enough to add a commit to bring these in line with the rest.

File: nixos/modules/services/networking/wireguard.nix [hunk 2]
```diff
@@ -118,23 +118,52 @@ let
...
+          This can be useful for setup where all traffic goes through the
+          wireguard tunnel, because the wireguard packets need to be routed
+          differently.
+        '';
+      };
+
+      mtu = mkOption {
+        default = null;
+        type = with types; nullOr int;
+        example = "1280";
```
Comment (fpletz) [inline nixos/modules/services/networking/wireguard.nix:160]:
```suggestion
        example = 1280;
```

Should be an int.

**Review by fpletz — COMMENTED** [review]
Small nitpick but otherwise looks good. Would be great if you could squash your formatting commit into the others or make just one commit from the code changes and the changelog entries.

File: nixos/modules/services/networking/wireguard.nix (hunk 2 above)
Comment (maifel-maifel) [reply nixos/modules/services/networking/wireguard.nix:160]: the formatting commit changes documentation of already existing options. if it does not belong here, I can just drop it.

**Review by jian-lin — COMMENTED** [review]
I don't think you should change existing description's format in this pr.

**Comment by maifel-maifel** [comment]
thanks for finding these, I applied them.

File: nixos/modules/services/networking/wireguard.nix [hunk 3]
```diff
@@ -406,6 +434,7 @@ let
           ${concatStringsSep " " (
             [ ''${wg} set "${name}" private-key "${privKey}"'' ]
             ++ optional (values.listenPort != null) ''listen-port "${toString values.listenPort}"''
+            ++ optional (values.fwMark != null) ''fwmark "${toString values.fwMark}"''
```
Comment (jian-lin) [inline nixos/modules/services/networking/wireguard.nix:437]:
```suggestion
            ++ optional (values.fwMark != null) ''fwmark "${values.fwMark}"''
```

File: nixos/modules/services/networking/wireguard.nix (hunk 2 above)
Comment (jian-lin) [reply nixos/modules/services/networking/wireguard.nix:158]: Do you forget to change this?

File: nixos/modules/services/networking/wireguard.nix (hunk 3 above)
Comment (jian-lin) [reply nixos/modules/services/networking/wireguard.nix:437]: Since its type is `str`, `toString` is not needed.

**Comment by maifel-maifel** [comment]
yes, I did forget that x.x
thanks for spotting that, the changes are applied now.

**Review by jian-lin — APPROVED** [review]

**Pull request merged**